samza-test/src/test/java/org/apache/samza/test/operator/StreamApplicationIntegrationTestHarness.java

```diff
@@ -0,0 +1,245 @@
+   * determined by {@link #NUM_EMPTY_POLLS}
+   *
+   * @param topics the list of topics to consume from
+   * @param threshold the number of messages to consume
+   * @return the list of {@link ConsumerRecord}s whose size can be atmost {@param threshold}
+   */
+  public List<ConsumerRecord<String, String>> getMessages(Collection<String> topics, int threshold) {
+    int emptyPollCount = 0;
+    List<ConsumerRecord<String, String>> recordList = new ArrayList<>();
+    consumer.subscribe(topics);
+
+    while (emptyPollCount < NUM_EMPTY_POLLS && recordList.size() < threshold) {
+      ConsumerRecords<String, String> records = consumer.poll(POLL_TIMEOUT_MS.toMillis());
+      if (!records.isEmpty()) {
+        Iterator<ConsumerRecord<String, String>> iterator = records.iterator();
+        while (iterator.hasNext() && recordList.size() < threshold) {
+          ConsumerRecord record = iterator.next();
+          recordList.add(record);
+        }
+      } else {
+        emptyPollCount++;
+      }
+    }
+    return recordList;
+  }
+
+  /**
+   * Executes the provided {@link StreamApplication} as a {@link org.apache.samza.job.local.ThreadJob}. The
+   * {@link StreamApplication} runs in its own separate thread.
+   *
+   * @param app the application to run
+   * @param appName the name of the application
+   * @param overriddenConfigs configs to override
+   */
+  public void runApplication(StreamApplication app, String appName, Config overriddenConfigs) {
+
+    Map<String, String> configs = new HashMap<>();
+    configs.put("job.factory.class", "org.apache.samza.job.local.ThreadJobFactory");
+    configs.put("job.name", appName);
+    configs.put("app.class", app.getClass().getCanonicalName());
+    configs.put("serializers.registry.json.class", "org.apache.samza.serializers.JsonSerdeFactory");
+    configs.put("serializers.registry.string.class", "org.apache.samza.serializers.StringSerdeFactory");
+    configs.put("systems.kafka.samza.factory", "org.apache.samza.system.kafka.KafkaSystemFactory");
+    configs.put("systems.kafka.consumer.zookeeper.connect", zkConnect());
+    configs.put("systems.kafka.producer.bootstrap.servers", bootstrapUrl());
+    configs.put("systems.kafka.samza.key.serde", "string");
+    configs.put("systems.kafka.samza.msg.serde", "string");
+    configs.put("systems.kafka.samza.offset.default", "oldest");
+    configs.put("job.coordinator.system", "kafka");
+    configs.put("job.default.system", "kafka");
+    configs.put("job.coordinator.replication.factor", "1");
+    configs.put("task.window.ms", "1000");
+
+    if (overriddenConfigs != null) {
+      configs.putAll(overriddenConfigs);
+    }
+
+    ApplicationRunner runner = ApplicationRunner.fromConfig(new MapConfig(configs));
+    runner.run(app);
+  }
+
+  /**
+   * Shutdown and clear Zookeeper and Kafka broker state.
+   */
+  @Override
+  public void tearDown() {
+    super.tearDown();
```

Comment:
I think this also needs to runner.stop() (or equivalent) to stop the application. If that's not implemented yet,
 (check with @xinyuiscool or @jmakes for the schedule) maybe add a stub or comment?